### PR TITLE
Order book for Uniswap V2 AMMs

### PR DESCRIPTION
**Pull request by faurdent**

### CODE REVIEW

**Comment by faurdent** [comment]
To test contents:
cd data_handler
poetry install
poetry shell
python ./order_books_tests.py

**Comment by lukaspetrasek** [comment]
@faurdent Replied on TG.

File: data_handler/handlers/order_books/ekubo/histogram.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+import matplotlib.pyplot as plt
+from typing import List
+
+from web_app.order_books.ekubo.main import EkuboOrderBook
+
+TOKEN_A = "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7"  # ETH
+TOKEN_B = "0x53c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8"  # USDC
+
+
+def fetch_order_book_and_current_price(token_a: str, token_b: str) -> tuple:
```
Comment (lukaspetrasek) [inline data_handler/handlers/order_books/ekubo/histogram.py:11]: Pls add types of individual items in the tuple. Holds for all other occurrences, and also for `dict`, `list`, `set`, etc.

File: data_handler/handlers/order_books/uniswap_v2/jediswap.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import asyncio
```
Comment (lukaspetrasek) [inline data_handler/handlers/order_books/uniswap_v2/jediswap.py:1]: This code is experimental? Does it work? Feels like `JediSwap` is covered in the `main`.

**Review by lukaspetrasek — CHANGES_REQUESTED** [review]

File: data_handler/handlers/order_books/uniswap_v2/main.py [hunk 3]
```diff
@@ -85,6 +85,24 @@ def tick_to_price(self, tick: Decimal) -> Decimal:
...
+    def calculate_token_amount_price_change(
+            self, current_price: Decimal, price_change_ratio: Decimal
+    ) -> Decimal:
+        """
+        Calculate amounts of the token_a required to change the price by the given ratio.
+        :param current_price: Decimal - The current pair price.
+        :param price_change_ratio: Decimal - The price change ratio.
+        :return: Decimal - Quantity that can be traded without moving price outside acceptable bound.
+        """
+        min_price = current_price - (current_price * price_change_ratio)
```
Comment (lukaspetrasek) [inline data_handler/handlers/order_books/uniswap_v2/main.py:97]:
```suggestion
        min_price = (decimal.Decimall("1") - price_change_ratio) * current_price
```

File: data_handler/handlers/order_books/uniswap_v2/main.py [hunk 4]
```diff
@@ -94,3 +112,4 @@ def tick_to_price(self, tick: Decimal) -> Decimal:
     order_book = UniswapV2OrderBook(token_0, token_1)
     order_book.fetch_price_and_liquidity()
     print(order_book.get_order_book(), "\n")
+    token_amount = order_book.calculate_token_amount_price_change(Decimal("3750"), Decimal(0.05))
```
Comment (lukaspetrasek) [inline data_handler/handlers/order_books/uniswap_v2/main.py:115]: The current price should be derived from the orderbook. It's the mid price between the best bid and ask.

**Review by lukaspetrasek — CHANGES_REQUESTED** [review]

File: data_handler/celery_app/celery_conf.py [hunk 5]
```diff
@@ -40,13 +43,19 @@
...
+    },
 }
+
 from celery_app.tasks import (
     # run_loan_states_computation_for_hashtack_v0,
     # run_loan_states_computation_for_hashtack_v1,
     run_loan_states_computation_for_nostra_alpha,
     # run_loan_states_computation_for_nostra_mainnet,
     # run_loan_states_computation_for_zklend,
+    uniswap_v2_order_book,
```
[the remark this answers is not in the document]
Comment (djeck1432) [reply data_handler/celery_app/celery_conf.py:58]: these tasks should be run first manually and then could be uncommented

**Comment by lukaspetrasek** [comment]
Hi @faurdent  ,

We'll be merging this PR so that it counts towards the OD Hackaton. You will receive a partial reward, but we'll be happy if you open a new PR and finish the task!

**Pull request merged**